File: dp.js
// Question 70
// Climbing Stairs

let climbStairs = (n) => {
  // fibinaci algo
  let prev = 0;
  let cur = 1;
  let temp;

  for (let i = 0; i < n; i++) {
    temp = prev;
    prev = cur;
    cur += temp;
  }
  return cur;
};

let climbStairs1 = (n, memo = { 1: 1, 2: 2 }) => {
  // using memoization and recursion
  if (memo[n] !== undefined) return memo[n];
  memo[n] = climbStairs(n - 1, memo) + climbStairs(n - 2, memo);
  return memo[n];
};

// Q198
// house robber

var rob2 = function (nums) {
  let memo = [];
  memo[0] = 0;
  memo[1] = nums[0];

  for (let i = 1; i < nums.length; i++) {
    memo[i + 1] = Math.max(memo[i], memo[i - 1] + nums[i]);
  }

  return memo[nums.length];
};

// Q322
// // Coin Change

const coinChange = (coins, amount) => {
  // dp[i] represents the least amount of coins that can make the value equals to the i
  const dp = Array(amount + 1).fill(Infinity);
  dp[0] = 0;
  for (let i = 1; i <= amount; i++) {
    for (const coin of coins) {
      if (i - coin >= 0) {
        dp[i] = Math.min(dp[i], dp[i - coin] + 1);
      }
    }
  }
  return dp[amount] === Infinity ? -1 : dp[amount];
};

// Q300
// Longest incraseing subsquence

var lengthOfLIS = function (nums) {
  const dp = Array.from(nums, () => 1);

  for (let i = 1; i < nums.length; i++) {
    for (let j = 0; j < i; j++) {
      if (nums[i] > nums[j]) {
        dp[i] = Math.max(dp[i], dp[j] + 1);
      }
    }
  }

  return Math.max(...dp);
};

// Q1143
// Longest common subsquence

let lcs = (t1, t2) => {
  const dp = Array(t1.length + 1)
    .fill(0)
    .map(() => Array(t2.length + 1).fill(0));

  for (let i = 1; i <= text1.length; i++) {
    for (let j = 1; j <= text2.length; j++) {
      if (text1[i - 1] === text2[j - 1]) {
        dp[i][j] = 1 + dp[i - 1][j - 1];
      } else {
        dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
      }
    }
  }
  return dp[0][0];
};

// Q62

// brute force recursive
const uniquePaths = (m, n) => {
  return helper(m, n, 1, 1);
};

const helper = (m, n, row, col) => {
  if (row === m && col === n) return 1;
  if (row > m || col > n) return 0;

  const pathsRight = helper(m, n, row, col + 1);
  const pathsDown = helper(m, n, row + 1, col);

  return pathsRight + pathsDown;
};

// Q 139
// Word Break

const wordBreak = (s, wordDict) => {
  if (wordDict == null || wordDict.length === 0) return false;

  const set = new Set(wordDict);
  const dp = Array(s.length + 1).fill(false);
  dp[0] = true;

  for (let end = 1; end <= s.length; end++) {
    for (let start = 0; start < end; start++) {
      const w = s.slice(start, end);
      if (dp[start] === true && set.has(w)) {
        dp[end] = true;
        break;
      }
    }
  }
  return dp[s.length];
};

//Q337
// Combination sum IV

// TC = O(mn); SC = O(m)
var combinationSum4 = function (nums, target) {
  const dp = Array(target + 1).fill(0);
  dp[0] = 1;

  // for each target sum, find the number of combinations possible
  for (let i = 1; i <= target; ++i) {
    nums.forEach((num) => {
      if (i >= num) {
        dp[i] += dp[i - num];
      }
    });
  }
  return dp[target];
};

// Q213

// the idea is to get two dp array one with the first house and one with the last house
var rob2 = function (nums) {
  let n = nums.length;
  if (n === 0) return 0;
  if (n === 1) return nums[0];
  if (n === 2) return Math.max(nums[0], nums[1]);

  let dp1 = new Array(n);
  let dp2 = new Array(n);

  computeResult(0, n - 2, dp1, nums);
  computeResult(1, n - 1, dp2, nums);

  function computeResult(i, n, dp, nums) {
    dp[i] = nums[i];
    dp[i + 1] = Math.max(dp[i], nums[i + 1]);

    for (let j = i + 2; j <= n; j++) {
      dp[j] = Math.max(dp[j - 1], dp[j - 2] + nums[j]);
    }
  }
  return Math.max(dp1[n - 2], dp2[n - 1]);
};

// Q91
// decode ways

const numDecodings = (s) => {
  const dp = Array(s.length + 1).fill(0);
  dp[0] = 1;
  for (let i = 1; i <= s.length; i++) {
    const one = Number(s.slice(i - 1, i)); // '0' to '9'
    const two = Number(s.slice(i - 2, i)); // '00' to '99'
    if (one !== 0) dp[i] += dp[i - 1]; // We want [1, 9]
    if (10 <= two && two <= 26) dp[i] += dp[i - 2]; // We want [10, 26]
  }
  return dp[s.length];
};

// DP
var uniquePaths = function (m, n) {
  let dp = new Array(m).fill(0).map(() => new Array(n));
  for (let row = 0; row < m; row++) {
    for (let col = 0; col < n; col++) {
      if (row === 0 || col === 0) {
        dp[row][col] = 1;
      } else {
        dp[row][col] = dp[row - 1][col] + dp[row][col - 1];
      }
    }
  }
  return dp[m - 1][n - 1];
  // T.C: O(M*N)
  // S.C: O(M*N)
};

// Q55
// jump game

var canJump = function (nums) {
  let max = 0;

  for (let i = 0; i < nums.length; i++) {
    if (i > max) return false;

    if (i + nums[i] >= nums.length - 1) return true;

    max = Math.max(max, i + nums[i]);
  }
};

File: dp.py
# Q70
# climbing stairs 

# Top down - TLE
def climbStairs1(self, n):
    if n == 1:
        return 1
    if n == 2:
        return 2
    return self.climbStairs(n-1)+self.climbStairs(n-2)
 
# Bottom up, O(n) space
def climbStairs2(self, n):
    if n == 1:
        return 1
    res = [0 for i in range(n)]
    res[0], res[1] = 1, 2
    for i in range(2, n):
        res[i] = res[i-1] + res[i-2]
    return res[-1]

# Bottom up, constant space
def climbStairs3(self, n):
    if n == 1:
        return 1
    a, b = 1, 2
    for i in range(2, n):
        tmp = b
        b = a+b
        a = tmp
    return b
    
# Top down + memorization (list)
def climbStairs4(self, n):
    if n == 1:
        return 1
    dic = [-1 for i in range(n)]
    dic[0], dic[1] = 1, 2
    return self.helper(n-1, dic)
    
def helper(self, n, dic):
    if dic[n] < 0:
        dic[n] = self.helper(n-1, dic)+self.helper(n-2, dic)
    return dic[n]
    
# Top down + memorization (dictionary)  
def __init__(self):
    self.dic = {1:1, 2:2}
    
def climbStairs(self, n):
    if n not in self.dic:
        self.dic[n] = self.climbStairs(n-1) + self.climbStairs(n-2)
    return self.dic[n]

# Q198
# climbing stairs 

def houseRobber(nums):
  rob1, rob2 = 0,0

  for i in nums:
    temp = max(i+rob1, rob2)
    rob1 = rob2
    rob2 = temp
  return rob2

# Q198
# climbing stairs 

# # DP
def coinChange(coins,amount):
  dp = [amount+1] * (amount+1)
  dp[0]=0

  for i in range(1, amount + 1):
    for c in coins:
      if i - c >= 0:
        dp[i] = min(dp[i], 1+dp[i-c])
  return dp[amount] if dp[amount] != amount+1 else -1

# Q300
# longest increasing sub

# DP bottom up
def lis(nums):
  dp = [1] * len(nums)

  for i in range(1, len(nums)):
    for j in range(i):
      if nums[i] > nums[j]:
        dp[i] = max(dp[i], dp[j]+1)
  
  return max(dp)

  # Brute-force recursion (TLE): time O(2^n) space O(n)
def lengthOfLIS(self, nums) -> int:
	def max_lis(idx, cur_max):
		if idx == len(nums):
			return 0
		if nums[idx] > cur_max:
			return max(1 + max_lis(idx + 1, nums[idx]), max_lis(idx + 1, cur_max))
		return max_lis(idx + 1, cur_max)
	return max_lis(0, float('-inf'))

# Recursion with memoization (DP): time O(n^2) space O(n^2)
from collections import defaultdict
def lengthOfLIS(self, nums) -> int:
	cache = defaultdict(dict) # 2D cache of prev_max_idx & cur_idx
	nums.append(float('-inf'))
	def max_lis(idx, prev_max_idx):
		if idx == len(nums) - 1:
			return 0
		if prev_max_idx not in cache or idx not in cache[prev_max_idx]:
			if nums[idx] > nums[prev_max_idx]:
				cache[prev_max_idx][idx] = max(1 + max_lis(idx + 1, idx), max_lis(idx + 1, prev_max_idx))
			else:
				cache[prev_max_idx][idx] = max_lis(idx + 1, prev_max_idx)
		return cache[prev_max_idx][idx]
	return max_lis(0, -1)

# Bottom-up DP + binary-search: time O(nlogn) space O(n)
def lengthOfLIS(self, nums) -> int:
	if not nums:
		return 0
	dp = [nums[0]]
	len_dp = 1
	for i in range(1, len(nums)):
		left, right = 0, len(dp) - 1
		while left < right:
			mid = (left + right) // 2
			if dp[mid] < nums[i]:
				left = mid + 1
			else:
				right = mid
		if dp[left] < nums[i]:
			dp.append(nums[i])
			len_dp += 1
		else:
			dp[left] = nums[i]
	return len_dp

# Q1143
# longest common sub

# 2d grid 
def lcs(text1,text2):
  dp = [[0 for j in range(len(text2)+1)] for i in range(len(text1)+1)] 
    
  for i in range(len(text1)-1, -1, -1): 
    for j in range(len(text2)-1, -1, -1):
        if text1[i]==text2[j]:
          dp[i][j] = 1 + dp[i+1][j+1]
        else:
          dp[i][j] = max(dp[i][j+1], dp[i+1][j])
  return dp[0][0]

# Q139
# word break

def wordBreak(s, wordDict):
  dp = [False] *(len(s)+1)
  dp[len(s)]= True

  for i in range(len(s)-1,-1,-1):
    for w in wordDict:
      if(i+len(w)) <= len(s) and s[i:i+len(w)] == w:
        dp[i] = dp[i+len(w)]
      if dp[i]:
        break
  return dp[0]

# Q139
# comb sum

def combSum4(nums,target):
  dp = {0:1}

  for total in range(1, target+1):
    dp[total] = 0
    for n in nums:
      dp[total] += dp.get(total-n,0)
  return dp[target]

# Q213
# house robber II

def houseRobber2(self,nums):
  # apply the function to first without last and last without first
  return max(nums[0], helper(nums[1:]),helper(nums[:-1]))

def helper(self,nums):
  rob1, rob2 = 0,0
  for i in nums:
    temp = max(i+rob1, rob2)
    rob1 = rob2
    rob2 = temp
  return rob2

# Q139
# decode ways

def decode(s):
  dp={ len(s):1}

  for i in range(len(s) -1, -1, -1):
    if s[i]== '0':
      dp[i]= 0
    else: 
      dp[i] = dp[i+1]
    
    if (i+1 < len(s) and (s[i] == '1' or s[i] == '2' and s[i+1] in '0123456')):
      dp[i] += dp[i+2]
  return dp[0]


# Q139
# unique paths

def uniquePaths(m,n):
  row = [1]*n

  for i in range(m-1):
    newRow = [1]*m
    for j in range(n-2-1,-1,-1):
      newRow[j] = newRow[j+1] +row[i]
    row=newRow
  return row[0]

# Q139
# jump game

def jumpGame(nums):
  goal = len(nums) -1

  for i in range(len(nums) -1, -1, -1):
    if i + nums[i] >= goal:
      goal= int(nums[i])
  return True if goal == 0 else False